registry/auth/openshift/accessController.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/docker/distribution/registry/auth"
+	"golang.org/x/net/context"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+)
+
+type accessController struct {
+	openshiftAddr       string
+	openshiftApiVersion string
+	client              *http.Client
+}
+
+type openshiftToken struct {
+	namespace   string
+	imageName   string
+	verb        string
+	bearerToken string
+}
+
+var _ auth.AccessController = &accessController{}
+
+func (ac *accessController) verifyOpenShiftToken(token openshiftToken) error {
+	url := fmt.Sprintf("https://%s/subjectAccessReviews?namespace=%s", ac.openshiftAddr, token.namespace)
+	sar := map[string]string{
+		"kind":         "SubjectAccessReview",
+		"apiVersion":   ac.openshiftApiVersion,
+		"verb":         token.verb,
+		"resource":     "imageRepositories",
+		"resourceName": token.imageName,
+	}
+
+	body, err := json.Marshal(sar)
+	if err != nil {
+		return fmt.Errorf("Error marshaling SubjectAccessReview: %s", sar)
+	}
+	req, err := http.NewRequest("POST", url, strings.NewReader(string(body)))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenShift request: %s", err)
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token.bearerToken))
+
+	resp, err := ac.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("Error querying OpenShift for SubjectAccessReview: %s", err)
+	}
+
+	defer resp.Body.Close()
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("Error reading OpenShift SubjectAccessReview response body: %s", err)
+	}
+
+	var accessResponse authorizationapi.SubjectAccessReviewResponse
+	err = json.Unmarshal(respBody, &accessResponse)
+	if err != nil {
+		return fmt.Errorf("Error parsing OpenShift SubjectAccessReviewResponse: %s", err)
+	}
+	if !accessResponse.Allowed {
+		return fmt.Errorf("OpenShift access denied: %s", accessResponse.Reason)
+	}
+	return nil
+}
+
+func (ac *accessController) Authorized(ctx context.Context, accessRecords ...auth.Access) (context.Context, error) {
+	req, err := ctxu.GetRequest(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(accessRecords) == 0 {
+		return ctx, nil
+	}
+	for _, access := range accessRecords {
+		//fmt.Printf("%s:%s:%s", access.Resource.Type, access.Resource.Name, access.Action)
+
+		if access.Resource.Type != "repository" {
+			continue
+		}
+		repoParts := strings.Split(access.Resource.Name, "/")
+		if len(repoParts) != 2 {
+			return nil, fmt.Errorf("Repository namespace not found for resource: %s", access.Resource.Name)
+		}
+		var token openshiftToken
+		token.namespace, token.imageName = repoParts[0], repoParts[1]
+
+		if access.Action == "push" {
+			token.verb = "create"
+		} else if access.Action == "pull" {
+			token.verb = "get"
+		} else {
+			return nil, fmt.Errorf("Unkown action: %s", access.Action)
+		}
+
+		authorization := req.Header.Get("Authorization")
+		if len(authorization) == 0 {
+			return nil, fmt.Errorf(`"Authorization" header not set`)
```

Comment:
Instead of doing this, you'll need to create a challenge to send back to the client (Docker) so it knows it needs to authenticate. See https://github.com/docker/distribution/blob/master/registry/auth/silly/access.go#L52-L66 for an example.

An example challenge implementation is here: https://github.com/docker/distribution/blob/master/registry/auth/silly/access.go#L72-L91. Note, however, you'll need to send back a "basic" challenge and not a "bearer" one.